aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+// TODO:  Remove this dependency once callBoxed is possible for all ops
+#include <torch/csrc/jit/operator.h>
+
+#include <iostream>
+#include <exception>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Imitate Apex and cache some of the casts to streamline parameter reuse.
+// Our heuristic is to cache FP16 casts of FP32 model weights (see cached_cast below).
+//
+// After discussion with @ezyang, the cache uses the following structure:
+// The key is the source tensor's TensorImpl*, a proxy for a Tensor uuid that's unchanged across shallow copies.
+// The value is a tuple of two Tensors:  the source tensor and the casted tensor.  Both are stored as full
+// Tensors to ensure their TensorImpls stay alive if other references to them die.
+//
+// We must keep the source tensor alive because if the source tensor were deallocated, another random Tensor
+// could be allocated whose TensorImpl* happened to have the same value.  This TensorImpl* would then mistakenly
+// hit in cache, which would be a nasty bug (rare, intermittent, unpredictable).
+//
+// Since our heuristic caches casts for model weights only, the source Tensors should always stay alive anyway,
+// because the model stores them explicitly.  So in the common case, storing a live reference to the
+// source tensor in cached_casts is unnecessary but also does no harm (does not increase memory use).
+//
+// When the autocast context manager exits, which should occur at the end of each forward pass, it calls
+// clear_cache to ensure cached Tensors don't leak outside the autocasting region.
+thread_local std::unordered_map<TensorImpl*, std::tuple<Tensor, Tensor>> cached_casts;
+}
+
+void clear_cache() {
+  cached_casts.clear();
+}
+
+enum class CastPolicy : uint8_t {
+  fp16 = 0, // Cast all inputs to fp16 before running the op.
+  fp16_with_tensorlist, // Separate from CastPolicy::fp16 due to lifetime concerns of data
+                        // viewed by TensorLists.  Maybe some C++ guru can help me unify them.
+  fp32, // Cast all inputs to fp32 before running the op.
+  fp32_set_opt_dtype, // fp32_set_opt_dtype and fp32_append_dtype handle functions that should
+  fp32_append_dtype, // output in fp32 and support an output dtype flag.  The policy is:
```

Comment:
nit: indentation of the continued lines don't mactch

Reply:
OK, so what's the difference between `set_opt` and `append`?

Reply:
Some functions already take an optional output dtype, like softmax [`- func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor`](https://github.com/pytorch/pytorch/blob/4980eee055a0d97a92bb4568328c47b0c14f20b2/aten/src/ATen/native/native_functions.yaml#L2499).  For these, I don't need a preceding cast.  My wrappers can just set the right output type if it's not already set.  That's the set_opt case.

Some functions (like [norm](https://github.com/pytorch/pytorch/blob/4980eee055a0d97a92bb4568328c47b0c14f20b2/aten/src/ATen/native/native_functions.yaml#L3033-L3038)) don't take an optional output dtype, but do have a different overload that takes a non-optional output dtype.    For these, I intercept calls of the non-dtype overload, append the right dtype to the arglist, and forward to the dtype-aware overload.
```
# overload with dtype, that I can redispatch to
- func: norm.ScalarOpt_dtype(Tensor self, Scalar? p, *, ScalarType dtype) -> Tensor
  variants: function, method

# overload without dtype, that I can intercept:
- func: norm.Scalar(Tensor self, Scalar p=2) -> Tensor
  use_c10_dispatcher: full
  variants: function, method
```
That's the append_dtype case.

The wrapper code required for both cases is simple, but divergent, so I consider them distinct policies (with distinct WrapFunction_ specializations).

Reply:
Updated comments with a hopefully improved explanation.